### ISSUE TITLE
Add find/search/where/single/firstWhere method

### ISSUE THREAD

**Issue by Radiergummi** [loophp/collection#198]
Currently there is no method to return the first collection item that passes a given predicate; this would be a shorthand for `filter($callable)->first()`, but is useful nevertheless.

- Laravel has [`firstWhere` (for values)](https://laravel.com/docs/8.x/collections#method-first-where) or [`search` (for keys)](https://laravel.com/docs/8.x/collections#method-search)
- Knapsack has [`find`](http://dusankasan.github.io/Knapsack/#Operations-find)
- nikic/iter has [`search`](https://github.com/nikic/iter)
- Ruby has [`find_index`](https://ruby-doc.org/core-2.7.0/Array.html#method-i-find_index)

It would be nice to have a method to quickly search for an item. I'd propose a syntax like the following:
```php
/**
 * @template T
 */
interface Searchable
{
    /**
     * @return T|null
     */
    public function search(callable ...$callbacks): mixed;
}
```

**Comment by drupol** [comment]
Hey hi !

Good idea! And it could be quickly implemented.

If I understand correctly, your search would be based on the key(TKey) and then it would return the associated value (T) if found? 

Question: I see that the proposed return type is `T|null`. Which means that it can return any type or `null`. What would be the behavior of that operation knowing that the collection can contains `null` values then?

**Comment by Radiergummi** [comment]
> If I understand correctly, your search would be based on the key(TKey) and then it would return the associated value (T) if found?

Ha, should've specified the callback. I intended to search for a value, so I thought of `($value: T, $key: TKey) => boolean`. It would return true for any matching value.

> Question: I see that the proposed return type is T|null. Which means that it can return any type or null. What would be the behavior of that operation knowing that the collection can contains null values then?

The operation would return the first value for which the callback returns true, and null if it never returned true, ie. there is no matching value in the collection. To counter the Problem of null values in Zheng collection, I would usually include a default argument, but that doesn't work for the variadic callbacks arg... Tricky. I wouldn't want to throw, but it looks like the only solution. Do you have any idea?

**Comment by drupol** [comment]
> Ha, should've specified the callback. I intended to search for a value, so I thought of `($value: T, $key: TKey) => boolean`. It would return true for any matching value.

Ok just like a filter callback.

> The operation would return the first value for which the callback returns true, and null if it never returned true, ie. there is no matching value in the collection. To counter the Problem of null values in Zheng collection, I would usually include a default argument, but that doesn't work for the variadic callbacks arg... Tricky. I wouldn't want to throw, but it looks like the only solution. Do you have any idea?

Yes, keep in mind that `loophp/collection` is about iterations.

Then, we could simply do: 

```php
$callback = static fn($value, $key): bool => 'a' ===  $value;

$search = Collection::fromIterable($inputData)->filter($callback);

$result = $search->isEmpty() ? 'UserDefaultValueWhenNoResultIsFound' : $search->first();
```

This way, there is no need to create a new operation and the user MUST implements it's own default value when no result is found. This is in my opinion, the best solution here.

**Comment by AlexandruGG** [comment]
I think I would find this useful as well, as long as it's simply a shortcut for the operations that we'd currently do. Meaning these would be equivalent:


```PHP
// Currently-available
// With found value
$found = Collection::fromIterable([1, 2, 3])
    ->filter(static fn(int $val): bool => $val % 2 === 0)
    ->first()
    ->current(); 
// result: 2

// Without found value
$notFound = Collection::fromIterable([1, 3, 5])
    ->filter(static fn(int $val): bool => $val % 2 === 0)
    ->first()
    ->current(); 
// result: NULL

//---------------------------
// New Operation - `Find`
// With found value
$found = Collection::fromIterable([1, 2, 3])
    ->find(static fn(int $val): bool => $val % 2 === 0);
// result: 2
    
// Without found value
$notFound = Collection::fromIterable([1, 3, 5])
    ->find(static fn(int $val): bool => $val % 2 === 0);
// result: NULL
```

Is this what you were thinking @Radiergummi?
And what do you think about this @drupol? In this form it would be a simple pipe between filter, first, current.

**Comment by Radiergummi** [comment]
> Yes, keep in mind that loophp/collection is about iterations.

@drupol sure, you're right. I often use functions like this as exit points from a long iterative processing step, however, where it's useful to have a clear way to extract a single item from the collection: It just makes the code easier to read.

> Is this what you were thinking @Radiergummi?

Yep, exactly. This has the same problem of null values in the collection not being discernible from missing values, and I don't have a good solution to that... Maybe it's okay to stick with filter/first/current :)

**Comment by AlexandruGG** [comment]
> This has the same problem of null values in the collection not being discernible from missing values

I am not sure this is a problem really, since if you use filter+first+current you get the same result.

It would be incorrect to use this sequence of operations or this new `find` operation to determine whether you have a `null` value in your collection. There are other operations that could help achieve that, such as: `truthy, nullsy, contains(null)`

**Comment by drupol** [comment]
> And what do you think about this @drupol? In this form it would be a simple pipe between filter, first, current.

Sure, but there is still an issue => "This has the same problem of null values in the collection not being discernible from missing values"

>There are other operations that could help achieve that, such as: `truthy, nullsy, contains(null)`

I fully agree.

Another idea would be to provide such an operation having this exact signature:

```php
Collection::find(callable $predicate, mixed $valueIfPredicateIsNotMet);
```

Examples:

```php
$value = Collection::fromIterable(['a', 'b', 'c'])->find(static fn($value, $key) => $key === 4, 'foo'); // 'foo'

$value = Collection::fromIterable(['a', 'b', 'c'])->find(static fn($value, $key) => $key === 2, 'foo'); // 'c'
```

Let me know what you think.

**Comment by Radiergummi** [comment]
> Another idea would be to provide such an operation having this exact signature:
> ```php
> Collection::find(callable $predicate, mixed $valueIfPredicateIsNotMet);
> ```

That would be great, but break with the way callbacks are used in the other collection functions, so I wasn't sure whether you'd accept that.

**Comment by drupol** [comment]
Let's wait for @AlexandruGG 's input and then we'll see how to tackle this.

**Comment by AlexandruGG** [comment]
> > Another idea would be to provide such an operation having this exact signature:
> > ```
> > Collection::find(callable $predicate, mixed $valueIfPredicateIsNotMet);
> > ```
> 
> That would be great, but break with the way callbacks are used in the other collection functions, so I wasn't sure whether you'd accept that.

Indeed, if we want to keep the same structure, which we should because it's one of the core [principles](https://loophp-collection.readthedocs.io/en/stable/pages/principles.html#variadic-parameters) of the package, then the signature would have to be:

```PHP
Collection::find(mixed $valueIfPredicateIsNotMet, callable ...$predicates);
```

Users might find this signature a bit weird though.

> Sure, but there is still an issue => "This has the same problem of null values in the collection not being discernible from missing values"

I think this is not an issue as long as the documentation specifies that `find` is basically a combination of other operations, because the user would need to be familiar with how `current` works regardless. The only true way around it is to use a language where `null` doesn't exist 😄 . Or use something like [php-option](https://github.com/schmittjoh/php-option)

**Comment by drupol** [comment]
Sure for me it's fine !

**Comment by Radiergummi** [comment]
I underestimated the complexity of the code base a little, didn't think it would take quite as much insight to chain a bunch of calls together. I won't get finished today.

**Comment by drupol** [comment]
This feature is now shipped in version 6.0.0 ! Congrats for your first contribution @Radiergummi !
